diff -urpN zlib-1.2.11/CMakeLists.txt zlib-1.2.11.patch/CMakeLists.txt
--- zlib-1.2.11/CMakeLists.txt	2017-01-15 18:29:40.000000000 +0100
+++ zlib-1.2.11.patch/CMakeLists.txt	2019-10-09 10:38:35.002060195 +0200
@@ -1,10 +1,18 @@
 cmake_minimum_required(VERSION 2.4.4)
 set(CMAKE_ALLOW_LOOSE_LOOP_CONSTRUCTS ON)
 
+if(${CMAKE_VERSION} VERSION_GREATER "3.15.0")
+  if(POLICY CMP0091)
+    cmake_policy(SET CMP0091 NEW)
+  endif()
+endif()
+
 project(zlib C)
 
 set(VERSION "1.2.11")
 
+SET(MSVC_STATIC_RUNTIME_LIBRARY OFF CACHE BOOL "Choose either ON or OFF.")
+
 option(ASM686 "Enable building i686 assembly implementation")
 option(AMD64 "Enable building amd64 assembly implementation")
 
@@ -14,6 +22,14 @@ set(INSTALL_INC_DIR "${CMAKE_INSTALL_PRE
 set(INSTALL_MAN_DIR "${CMAKE_INSTALL_PREFIX}/share/man" CACHE PATH "Installation directory for manual pages")
 set(INSTALL_PKGCONFIG_DIR "${CMAKE_INSTALL_PREFIX}/share/pkgconfig" CACHE PATH "Installation directory for pkgconfig (.pc) files")
 
+if(MSVC)
+  if(NOT BUILD_SHARED_LIBS)
+	if(MSVC_STATIC_RUNTIME_LIBRARY)
+      set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
+	endif()
+  endif()
+endif()
+
 include(CheckTypeSize)
 include(CheckFunctionExists)
 include(CheckIncludeFile)
