diff --git a/OMCompiler/configure.ac b/OMCompiler/configure.ac
index af5db0678..24aefd29e 100644
--- a/OMCompiler/configure.ac
+++ b/OMCompiler/configure.ac
@@ -205,13 +205,7 @@ if echo $CC | grep -v ccc-analyzer; then # ccc-analyzer calls gcc first and then
     TRY_FLAGS="$TRY_FLAGS -fno-ipa-pure-const"
   fi
 
-  # GCC 7.x issues:
-  # - RUNPATH instead of RPATH, disable that via -Wl,--disable-new-dtags
-  # - sqrt from libcolamd is not found in -lm if -Wl,--no-as-needed is not specified
-  if $CC --version | grep -q gcc && $CC -dumpversion | grep -q "^7"; then
-    TRY_FLAGS="$TRY_FLAGS -Wl,--no-as-needed"
-    RPATH_EXTRA="-Wl,--disable-new-dtags"
-  fi
+  TRY_FLAGS_LD="$TRY_FLAGS_LD -Wl,--no-as-needed -Wl,--disable-new-dtags"
 
 fi
 
@@ -222,6 +216,15 @@ for flag in $TRY_FLAGS; do
   AC_TRY_LINK([], [return 0;], [CFLAGS="$OLD_CFLAGS $flag"],[CFLAGS="$OLD_CFLAGS"])
 done
 AC_MSG_RESULT([$CFLAGS])
+
+AC_MSG_CHECKING([for flags to append to LDFLAGS ($TRY_FLAGS_LD)])
+for flag in $TRY_FLAGS_LD; do
+  OLD_LDFLAGS="$LDFLAGS"
+  LDFLAGS="$flag -Werror"
+  AC_TRY_LINK([], [return 0;], [LDFLAGS="$OLD_LDFLAGS $flag"],[LDFLAGS="$OLD_LDFLAGS"])
+done
+AC_MSG_RESULT([$LDFLAGS])
+
 OLD_CFLAGS="$CFLAGS"
 for flag in -Wno-parentheses-equality -Wno-unused-variable; do
   CFLAGS="$RUNTIMECFLAGS $flag -Werror"
@@ -742,8 +745,8 @@ else
   LDFLAGS="$LDFLAGS -Wl,-rpath-link,$OMBUILDDIR/lib/$host_short/omc"
   RT_LDFLAGS_SHARED="-Wl,-rpath-link,$OMBUILDDIR/lib/$host_short/omc"
   OMCRUNTIME_SHARED_LDFLAGS="$RT_LDFLAGS $LIBLPSOLVE55 -lzmq $LIBUUID"
-  RPATH="$RPATH_EXTRA -Wl,-z,origin -Wl,-rpath,'\$\$ORIGIN/../lib/$host_short/omc' -Wl,-rpath,'\$\$ORIGIN'"
-  RPATH_QMAKE="$RPATH_EXTRA -Wl,-z,origin -Wl,-rpath,\\'\\\$\$ORIGIN/../lib/$host_short/omc\\' -Wl,-rpath,\\'\\\$\$ORIGIN\\'"
+  RPATH="-Wl,-z,origin -Wl,-rpath,'\$\$ORIGIN/../lib/$host_short/omc' -Wl,-rpath,'\$\$ORIGIN'"
+  RPATH_QMAKE="-Wl,-z,origin -Wl,-rpath,\\'\\\$\$ORIGIN/../lib/$host_short/omc\\' -Wl,-rpath,\\'\\\$\$ORIGIN\\'"
   OS_TARGET="linux"
   AR_SH="$AR -ru"
   BOOTSTRAP_STATIC=""
diff --git a/OMCompiler/3rdParty/SuiteSparse/CMakeLists.txt b/OMCompiler/3rdParty/SuiteSparse/CMakeLists.txt
index e3c398ca..5b221d4e 100644
--- a/OMCompiler/3rdParty/SuiteSparse/CMakeLists.txt
+++ b/OMCompiler/3rdParty/SuiteSparse/CMakeLists.txt
@@ -67,7 +67,9 @@ if(SS_DIR)
     list(APPEND FLAGS "DLONG")
     set_target_properties(colamd_object_l PROPERTIES COMPILE_DEFINITIONS ${FLAGS})
     add_library(colamd $<TARGET_OBJECTS:colamd_object> $<TARGET_OBJECTS:colamd_object_l>)
-
+    if (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUXX)
+      target_link_libraries(colamd m)
+    endif (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUXX)
     install(TARGETS colamd DESTINATION ${LIB_OMC})
     install(FILES ${COLAMD_DIR}/Include/colamd.h DESTINATION Include)
 
@@ -162,6 +164,9 @@ if(SS_DIR)
   set_target_properties(colamd_object PROPERTIES COMPILE_DEFINITIONS "DINT")
   set_target_properties(colamd_object_l PROPERTIES COMPILE_DEFINITIONS "DLONG")
   add_library(colamd $<TARGET_OBJECTS:colamd_object> $<TARGET_OBJECTS:colamd_object_l>)
+  if (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUXX)
+    target_link_libraries(colamd m)
+  endif (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUXX)
   install(TARGETS colamd DESTINATION ${LIB_OMC})
   install(FILES ${COLAMD_DIR}/Include/colamd.h DESTINATION include/omc/c/suitesparse/Include)
 
