diff -urpN sundials-6.3.0/include/kinsol/kinsol.h sundials-6.3.0_patched/include/kinsol/kinsol.h
--- sundials-6.3.0/include/kinsol/kinsol.h	2022-08-11 06:45:15.000000000 +0200
+++ sundials-6.3.0_patched/include/kinsol/kinsol.h	2023-11-22 09:55:03.402892533 +0100
@@ -132,6 +132,7 @@ SUNDIALS_EXPORT int KINSetFuncNormTol(vo
 SUNDIALS_EXPORT int KINSetScaledStepTol(void *kinmem, realtype scsteptol);
 SUNDIALS_EXPORT int KINSetConstraints(void *kinmem, N_Vector constraints);
 SUNDIALS_EXPORT int KINSetSysFunc(void *kinmem, KINSysFn func);
+SUNDIALS_EXPORT int KINSetInitialAdditionalTolerance(void *kinmem, realtype initialaddtol);
 
 /* Optional input functions for handling error/info/debug events */
 SUNDIALS_EXPORT int KINSetErrHandlerFn(void *kinmem, KINErrHandlerFn ehfun,
diff -urpN sundials-6.3.0/src/ida/ida.c sundials-6.3.0_patched/src/ida/ida.c
--- sundials-6.3.0/src/ida/ida.c	2022-08-11 06:41:29.000000000 +0200
+++ sundials-6.3.0_patched/src/ida/ida.c	2023-11-22 09:55:28.398981849 +0100
@@ -3329,8 +3329,10 @@ static int IDARcheck2(IDAMem IDA_mem)
   if (IDA_mem->ida_irfnd == 0) return(IDA_SUCCESS);
 
   (void) IDAGetSolution(IDA_mem, IDA_mem->ida_tlo, IDA_mem->ida_yy, IDA_mem->ida_yp);
-  retval = IDA_mem->ida_gfun(IDA_mem->ida_tlo, IDA_mem->ida_yy, IDA_mem->ida_yp,
-                             IDA_mem->ida_glo, IDA_mem->ida_user_data);
+  // use of temporary variable (trout,grout) to detect sign change of g
+  IDA_mem->ida_trout = IDA_mem->ida_tlo;
+  retval = IDA_mem->ida_gfun(IDA_mem->ida_trout, IDA_mem->ida_yy, IDA_mem->ida_yp,
+                             IDA_mem->ida_grout, IDA_mem->ida_user_data);
   IDA_mem->ida_nge++;
   if (retval != 0) return(IDA_RTFUNC_FAIL);
 
@@ -3339,7 +3341,7 @@ static int IDARcheck2(IDAMem IDA_mem)
     IDA_mem->ida_iroots[i] = 0;
   for (i = 0; i < IDA_mem->ida_nrtfn; i++) {
     if (!IDA_mem->ida_gactive[i]) continue;
-    if (SUNRabs(IDA_mem->ida_glo[i]) == ZERO) {
+    if (SUNRabs(IDA_mem->ida_grout[i]) == ZERO) {
       zroot = SUNTRUE;
       IDA_mem->ida_iroots[i] = 1;
     }
@@ -3372,6 +3374,10 @@ static int IDARcheck2(IDAMem IDA_mem)
       if (IDA_mem->ida_iroots[i] == 1) return(CLOSERT);
       zroot = SUNTRUE;
       IDA_mem->ida_iroots[i] = 1;
+    }
+    else if (IDA_mem->ida_grout[i]*IDA_mem->ida_glo[i] < ZERO) {
+      zroot = SUNTRUE;
+      IDA_mem->ida_iroots[i] = IDA_mem->ida_grout[i] > 0 ? -1 : 1;
     } else {
       if (IDA_mem->ida_iroots[i] == 1)
         IDA_mem->ida_glo[i] = IDA_mem->ida_ghi[i];
diff -urpN sundials-6.3.0/src/kinsol/kinsol.c sundials-6.3.0_patched/src/kinsol/kinsol.c
--- sundials-6.3.0/src/kinsol/kinsol.c	2022-08-11 06:45:15.000000000 +0200
+++ sundials-6.3.0_patched/src/kinsol/kinsol.c	2023-11-22 09:56:09.315128052 +0100
@@ -290,6 +290,7 @@ void *KINCreate(SUNContext sunctx)
   kin_mem->kin_sqrt_relfunc     = SUNRsqrt(uround);
   kin_mem->kin_scsteptol        = SUNRpowerR(uround,TWOTHIRDS);
   kin_mem->kin_fnormtol         = SUNRpowerR(uround,ONETHIRD);
+  kin_mem->kin_initialaddtol    = POINT1;
   kin_mem->kin_etaflag          = KIN_ETACHOICE1;
   kin_mem->kin_eta              = POINT1;     /* default for KIN_ETACONSTANT */
   kin_mem->kin_eta_alpha        = TWO;        /* default for KIN_ETACHOICE2  */
@@ -1467,7 +1468,7 @@ static int KINSolInit(KINMem kin_mem)
   }
 
   fmax = KINScFNorm(kin_mem, kin_mem->kin_fval, kin_mem->kin_fscale);
-  if (fmax <= (POINT01 * kin_mem->kin_fnormtol)) {
+  if (fmax <= (kin_mem->kin_initialaddtol*kin_mem->kin_fnormtol)) {
     kin_mem->kin_fnorm = N_VWL2Norm(kin_mem->kin_fval, kin_mem->kin_fscale);
     return(KIN_INITIAL_GUESS_OK);
   }
diff -urpN sundials-6.3.0/src/kinsol/kinsol_impl.h sundials-6.3.0_patched/src/kinsol/kinsol_impl.h
--- sundials-6.3.0/src/kinsol/kinsol_impl.h	2022-08-11 06:45:15.000000000 +0200
+++ sundials-6.3.0_patched/src/kinsol/kinsol_impl.h	2023-11-22 09:55:40.647025614 +0100
@@ -87,6 +87,7 @@ typedef struct KINMemRec {
   void *kin_user_data;         /* work space available to func routine         */
   realtype kin_fnormtol;       /* stopping tolerance on L2-norm of function
                                   value                                        */
+  realtype kin_initialaddtol;  /* stopping tolerance on initialization         */
   realtype kin_scsteptol;      /* scaled step length tolerance                 */
   int kin_globalstrategy;      /* choices are KIN_NONE, KIN_LINESEARCH
                                   KIN_PICARD and KIN_FP                        */
@@ -443,6 +444,7 @@ void KINInfoHandler(const char *module,
 #define MSG_BAD_MXNEWTSTEP     "Illegal mxnewtstep < 0."
 #define MSG_BAD_RELFUNC        "relfunc < 0 illegal."
 #define MSG_BAD_FNORMTOL       "fnormtol < 0 illegal."
+#define MSG_BAD_INITADDTOL     "initialaddtol < 0 illegal."
 #define MSG_BAD_SCSTEPTOL      "scsteptol < 0 illegal."
 #define MSG_BAD_MXNBCF         "mxbcf < 0 illegal."
 #define MSG_BAD_CONSTRAINTS    "Illegal values in constraints vector."
diff -urpN sundials-6.3.0/src/kinsol/kinsol_io.c sundials-6.3.0_patched/src/kinsol/kinsol_io.c
--- sundials-6.3.0/src/kinsol/kinsol_io.c	2022-08-11 06:45:15.000000000 +0200
+++ sundials-6.3.0_patched/src/kinsol/kinsol_io.c	2023-11-22 09:55:45.827044123 +0100
@@ -856,6 +856,37 @@ int KINSetFuncNormTol(void *kinmem, real
   return(KIN_SUCCESS);
 }
 
+// Function : KINSetInitialAdditionalTolerance
+
+int KINSetInitialAdditionalTolerance(void *kinmem, realtype initialaddtol)
+{
+  KINMem kin_mem;
+  realtype uround;
+
+  if (kinmem == NULL) {
+    KINProcessError(NULL, KIN_MEM_NULL, "KINSOL", "KINSetInitialAdditionalTolerance", MSG_NO_MEM);
+    return(KIN_MEM_NULL);
+  }
+
+  kin_mem = (KINMem) kinmem;
+
+  if (initialaddtol < ZERO) {
+    KINProcessError(NULL, KIN_ILL_INPUT, "KINSOL", "KINSetInitialAdditionalTolerance", MSG_BAD_INITADDTOL);
+    return(KIN_ILL_INPUT);
+  }
+
+  if (initialaddtol == ZERO) {
+    uround = kin_mem->kin_uround;
+    kin_mem->kin_initialaddtol = SUNRpowerR(uround,ONETHIRD);
+  } else {
+    kin_mem->kin_initialaddtol = initialaddtol;
+  }
+
+  return(KIN_SUCCESS);
+}
+
+// End KINSetInitialAdditionalTolerance
+
 /*
  * -----------------------------------------------------------------
  * Function : KINSetScaledStepTol
diff -urpN sundials-6.3.0/src/kinsol/kinsol_ls.c sundials-6.3.0_patched/src/kinsol/kinsol_ls.c
--- sundials-6.3.0/src/kinsol/kinsol_ls.c	2022-08-11 06:45:15.000000000 +0200
+++ sundials-6.3.0_patched/src/kinsol/kinsol_ls.c	2023-11-22 09:55:52.751068864 +0100
@@ -248,6 +248,8 @@ int KINSetJacFn(void *kinmem, KINLsJacFn
     kinls_mem->J_data = kin_mem;
   }
 
+  kinls_mem->nje = 0;
+
   return(KINLS_SUCCESS);
 }
 
