diff -urpN dlfcn-win32-1.2.0/CMakeLists.txt dlfcn-win32-1.2.0.patch/CMakeLists.txt
--- dlfcn-win32-1.2.0/CMakeLists.txt	2019-05-20 23:41:19.000000000 +0200
+++ dlfcn-win32-1.2.0.patch/CMakeLists.txt	2019-10-09 10:35:24.761910012 +0200
@@ -1,14 +1,29 @@
 cmake_minimum_required(VERSION 2.8)
 
+if(${CMAKE_VERSION} VERSION_GREATER "3.15.0")
+  if(POLICY CMP0091)
+    cmake_policy(SET CMP0091 NEW)
+  endif()
+endif()
+
 if (NOT DEFINED CMAKE_BUILD_TYPE)
   set (CMAKE_BUILD_TYPE Release CACHE STRING "Build type")
 endif ()
 
 project (dlfcn-win32 C)
 
+SET(MSVC_STATIC_RUNTIME_LIBRARY OFF CACHE BOOL "Choose either ON or OFF.")
 option(BUILD_SHARED_LIBS "shared/static libs" ON) 
 option(BUILD_TESTS "tests?" OFF)
 
+if(MSVC)
+  if(NOT BUILD_SHARED_LIBS)
+    if(MSVC_STATIC_RUNTIME_LIBRARY)
+      set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
+	endif()
+  endif()
+endif()
+
 set(headers dlfcn.h)
 set(sources dlfcn.c)
 
