diff -urpN boost_1_64_0/boost/process/environment.hpp boost_1_64_0.patch/boost/process/environment.hpp
--- boost_1_64_0/boost/process/environment.hpp	2017-04-17 04:22:23.000000000 +0200
+++ boost_1_64_0.patch/boost/process/environment.hpp	2020-02-19 14:52:11.574928297 +0100
@@ -669,7 +669,7 @@ inline std::vector<boost::filesystem::pa
 {
 #if defined(BOOST_WINDOWS_API)
     const ::boost::process::wnative_environment ne{};
-    typedef typename ::boost::process::wnative_environment::const_entry_type value_type;
+    typedef ::boost::process::wnative_environment::const_entry_type value_type;
     const auto id = L"PATH";
 #else
     const ::boost::process::native_environment ne{};
